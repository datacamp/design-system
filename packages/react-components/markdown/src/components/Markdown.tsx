import {
  Code,
  CodeBlock,
  Emphasis,
  Heading,
  Paragraph,
  Strong,
  Text,
} from '@datacamp/waffles-text';
import {
  HeadingElement,
  Size,
} from '@datacamp/waffles-text/types/components/Heading';
import React, { ReactElement } from 'react';
// eslint-disable-next-line import/no-named-default
import { default as ReactMarkdown } from 'react-markdown';

interface ChildrenRendererProps {
  children: ReactElement[];
}
interface CodeRendererProps {
  value: string;
}
interface HeadingRendererProps extends ChildrenRendererProps {
  level: number;
}
interface MarkdownProps {
  source: string;
}

const headingSizes: { [key: number]: number } = {
  1: 800,
  2: 700,
  3: 600,
  4: 500,
  5: 400,
  6: 300,
};

const renderers = {
  code: (props: CodeRendererProps) => {
    const { value } = props;
    return <CodeBlock>{value}</CodeBlock>;
  },
  emphasis: (props: ChildrenRendererProps) => {
    const { children } = props;
    const childrenString: string = children[0].props.children;
    return <Emphasis>{childrenString}</Emphasis>;
  },
  heading: (props: HeadingRendererProps) => {
    const { level, children } = props;
    const childrenString: string = children[0].props.children;
    return (
      <Heading
        as={`h${level}` as HeadingElement}
        size={headingSizes[level] as Size}
      >
        {childrenString}
      </Heading>
    );
  },
  inlineCode: (props: CodeRendererProps) => {
    const { value } = props;
    return <Code>{value}</Code>;
  },
  paragraph: (props: ChildrenRendererProps) => {
    const { children } = props;
    return <Paragraph>{children}</Paragraph>;
  },
  strong: (props: ChildrenRendererProps) => {
    const { children } = props;
    const childrenString: string = children[0].props.children;
    return <Strong>{childrenString}</Strong>;
  },
  text: (props: ChildrenRendererProps) => {
    const { children } = props;
    return <Text>{children}</Text>;
  },
};

const Markdown: React.FC<MarkdownProps> = props => {
  const { source } = props;
  return <ReactMarkdown renderers={renderers} source={source} />;
};

export default Markdown;
