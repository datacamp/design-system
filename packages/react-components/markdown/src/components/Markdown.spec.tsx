import fs from 'fs';

import { render } from '@testing-library/react';
import React from 'react';

import Markdown from './Markdown';

const fullExample: string = fs
  .readFileSync(`${__dirname}/../../test/full-example.md`)
  .toString();

describe('<Markdown />', () => {
  it('renders', () => {
    const { container } = render(<Markdown source={fullExample} />);
    expect(container).toMatchSnapshot();
  });
});
