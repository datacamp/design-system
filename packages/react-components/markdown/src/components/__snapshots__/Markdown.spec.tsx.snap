// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Markdown /> renders 1`] = `
.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  color: rgb(61,66,81);
  font-weight: 800;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 40px;
}

*:not(style) ~ .emotion-0 {
  margin-top: 24px;
}

.emotion-1 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  color: rgb(61,66,81);
  font-weight: 800;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 32px;
}

*:not(style) ~ .emotion-1 {
  margin-top: 16px;
}

.emotion-2 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  color: rgb(61,66,81);
  font-weight: 800;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 24px;
}

*:not(style) ~ .emotion-2 {
  margin-top: 16px;
}

.emotion-3 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  color: rgb(61,66,81);
  font-weight: 800;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 20px;
}

*:not(style) ~ .emotion-3 {
  margin-top: 16px;
}

.emotion-4 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  color: rgb(61,66,81);
  font-weight: 800;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 18px;
}

*:not(style) ~ .emotion-4 {
  margin-top: 16px;
}

.emotion-5 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  color: rgb(61,66,81);
  font-weight: 800;
  line-height: 1.25;
  margin: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-size: 16px;
}

*:not(style) ~ .emotion-5 {
  margin-top: 16px;
}

.emotion-6 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-weight: 800;
  line-height: 16px;
}

.emotion-8 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  margin: 0;
}

*:not(style) ~ .emotion-8 {
  margin-top: 8px;
}

.emotion-11 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-style: italic;
  font-weight: 400;
  line-height: 16px;
}

.emotion-29 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  list-style-position: inside;
  margin: 0;
  padding: 0;
}

.emotion-29 .emotion-29 {
  padding-left: 16px;
}

*:not(style) ~ .emotion-29 {
  margin-top: 16px;
}

.emotion-26 {
  line-height: 1.5em;
  margin-top: 8px;
}

.emotion-45 {
  border: 0;
  border-radius: 4px;
  color: rgb(61,66,81);
  font-family: monospace;
  font-family: Roboto Mono;
  font-size: 14px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.5;
  margin: 0;
  padding: 0;
  background-color: rgb(235,244,247);
  box-shadow: 0 0 0 1px #e6eaeb;
  margin-left: 2px;
  margin-right: 4px;
  padding-bottom: 0;
  padding-left: 2px;
  padding-right: 2px;
  padding-top: 0;
}

.emotion-48 {
  background-color: rgb(235,244,247);
  border: 0;
  border-radius: 4px;
  box-shadow: 0 0 0 1px #e6eaeb;
  line-height: 1.5;
  margin: 0;
  overflow: scroll;
  padding-bottom: 8px;
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
}

*:not(style) ~ .emotion-48 {
  margin-top: 12px;
}

.emotion-47 {
  border: 0;
  border-radius: 4px;
  color: rgb(61,66,81);
  font-family: monospace;
  font-family: Roboto Mono;
  font-size: 14px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 1.5;
  margin: 0;
  padding: 0;
}

<div>
  <h1
    class="emotion-0"
  >
    This is an h1 tag
  </h1>
  <h2
    class="emotion-1"
  >
    This is an h2 tag
  </h2>
  <h3
    class="emotion-2"
  >
    This is an h3 tag
  </h3>
  <h4
    class="emotion-3"
  >
    This is an h4 tag
  </h4>
  <h5
    class="emotion-4"
  >
    This is an h5 tag
  </h5>
  <h6
    class="emotion-5"
  >
    This is an h6 tag
  </h6>
  <h2
    class="emotion-1"
  >
    Heading with 
    <strong
      class="emotion-6"
    >
      emphasis
    </strong>
     inside
  </h2>
  <p
    class="emotion-8"
  >
    This is the first paragraph.
  </p>
  <p
    class="emotion-8"
  >
    This is a second paragraph.
  </p>
  <p
    class="emotion-8"
  >
    This paragraph should
    <br />
    have 2 lines.
  </p>
  <p
    class="emotion-8"
  >
    <em
      class="emotion-11"
    >
      This text will be italic
    </em>
  </p>
  <p
    class="emotion-8"
  >
    <em
      class="emotion-11"
    >
      This will also be italic
    </em>
  </p>
  <p
    class="emotion-8"
  >
    <strong
      class="emotion-6"
    >
      This text will be bold
    </strong>
  </p>
  <p
    class="emotion-8"
  >
    <strong
      class="emotion-6"
    >
      This will also be bold
    </strong>
  </p>
  <p
    class="emotion-8"
  >
    Only the word 
    <strong
      class="emotion-6"
    >
      strong
    </strong>
     should be strong.
  </p>
  <p
    class="emotion-8"
  >
    <em
      class="emotion-11"
    >
      You 
      <strong
        class="emotion-6"
      >
        can
      </strong>
       combine them
    </em>
  </p>
  <p
    class="emotion-8"
  >
    This paragraph has a 
    <del>
      strikthrough
    </del>
     in it.
  </p>
  <h3
    class="emotion-2"
  >
    Unordered list
  </h3>
  <ul
    class="emotion-29"
  >
    <li
      class="emotion-26"
    >
      <span
        id="1val-checkbox-description"
      >
        Item 1
      </span>
    </li>
    <li
      class="emotion-26"
    >
      <span
        id="2val-checkbox-description"
      >
        Item 2
        <ul
          class="emotion-29"
        >
          <li
            class="emotion-26"
          >
            <span
              id="3val-checkbox-description"
            >
              Item 2a
            </span>
          </li>
          <li
            class="emotion-26"
          >
            <span
              id="4val-checkbox-description"
            >
              Item 2b
            </span>
          </li>
        </ul>
      </span>
    </li>
  </ul>
  <h3
    class="emotion-2"
  >
    Ordered list
  </h3>
  <ol
    class="emotion-29"
  >
    <li
      class="emotion-26"
    >
      <span
        id="5val-checkbox-description"
      >
        Item 1
      </span>
    </li>
    <li
      class="emotion-26"
    >
      <span
        id="6val-checkbox-description"
      >
        Item 2
      </span>
    </li>
    <li
      class="emotion-26"
    >
      <span
        id="7val-checkbox-description"
      >
        Item 3
        <ol
          class="emotion-29"
        >
          <li
            class="emotion-26"
          >
            <span
              id="8val-checkbox-description"
            >
              Item 3a
            </span>
          </li>
          <li
            class="emotion-26"
          >
            <span
              id="9val-checkbox-description"
            >
              Item 3b
            </span>
          </li>
        </ol>
      </span>
    </li>
  </ol>
  <p
    class="emotion-8"
  >
    <img
      alt="This is the alt text"
      src="https://placehold.it/350x150"
    />
  </p>
  <p
    class="emotion-8"
  >
    <a
      href="http://github.com"
    >
      http://github.com
    </a>
     - automatic!
  </p>
  <p
    class="emotion-8"
  >
    <a
      href="http://github.com"
    >
      GitHub
    </a>
  </p>
  <p
    class="emotion-8"
  >
    As Kanye West said:
  </p>
  <blockquote>
    <p
      class="emotion-8"
    >
      We're living the future so
the present is our past.
    </p>
  </blockquote>
  <p
    class="emotion-8"
  >
    This paragraphs contains 
    <code
      class="emotion-45"
    >
      inline code
    </code>
    .
  </p>
  <pre
    class="emotion-48"
    tabindex="0"
  >
    <code
      class="emotion-47"
    >
      This is a code block
with 2 lines
  and 2 space indentation
    </code>
  </pre>
  <p
    class="emotion-8"
  >
    The following code block should highlight python code:
  </p>
  <pre
    class="emotion-48"
    tabindex="0"
  >
    <code
      class="emotion-47"
    >
      numbers = [1, 2, 3]
asStrings = [str(x) for x in stuff]
    </code>
  </pre>
  <p
    class="emotion-8"
  >
    The following code block should highlight javascript code:
  </p>
  <pre
    class="emotion-48"
    tabindex="0"
  >
    <code
      class="emotion-47"
    >
      var React = require('react');
var Markdown = require('react-markdown');

React.render(
  &lt;Markdown source="# Your markdown here" /&gt;,
  document.getElementById('content')
);
    </code>
  </pre>
  <ul
    class="emotion-29"
  >
    <li
      class="emotion-26"
    >
      <input
        aria-labelledby="10val-checkbox-description"
        checked=""
        disabled=""
        readonly=""
        type="checkbox"
      />
      <span
        id="10val-checkbox-description"
      >
        @mentions, #refs, 
        <a
          href="http://github.com"
        >
          links
        </a>
        , 
        <strong
          class="emotion-6"
        >
          formatting
        </strong>
        , and 
        &lt;del&gt;
        tags
        &lt;/del&gt;
         supported
      </span>
    </li>
    <li
      class="emotion-26"
    >
      <input
        aria-labelledby="11val-checkbox-description"
        checked=""
        disabled=""
        readonly=""
        type="checkbox"
      />
      <span
        id="11val-checkbox-description"
      >
        list syntax required (any unordered or ordered list supported)
      </span>
    </li>
    <li
      class="emotion-26"
    >
      <input
        aria-labelledby="12val-checkbox-description"
        checked=""
        disabled=""
        readonly=""
        type="checkbox"
      />
      <span
        id="12val-checkbox-description"
      >
        this is a complete item
      </span>
    </li>
    <li
      class="emotion-26"
    >
      <input
        aria-labelledby="13val-checkbox-description"
        disabled=""
        readonly=""
        type="checkbox"
      />
      <span
        id="13val-checkbox-description"
      >
        this is an incomplete item
      </span>
    </li>
  </ul>
  <table>
    <thead>
      <tr>
        <th>
          First Header
        </th>
        <th>
          Second Header
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Content from cell 1
        </td>
        <td>
          Content from cell 2
        </td>
      </tr>
      <tr>
        <td>
          Content in the first column
        </td>
        <td>
          Content in the second column
        </td>
      </tr>
    </tbody>
  </table>
  &lt;h1&gt;This is an h1&lt;br&gt;with HTML&lt;/h1&gt;
&lt;table&gt;
  &lt;tr&gt;
    &lt;td&gt;hi&lt;/td&gt;
    &lt;td&gt;there&lt;/td&gt;
    &lt;td&gt;its a table&lt;/td&gt;
  &lt;/tr&gt;
&lt;/table&gt;
  &lt;script&gt;alert("obviously you shouldn’t see this")&lt;/script&gt;
  &lt;style&gt;body{border: 40px dotted red !important;}&lt;/style&gt;
</div>
`;
